man/man1/atlc.1
.TH atlc 1 "atlc-3.0.1 29/3/02" "Dr. David Kirkby"
.ds n 5
.SH NAME
\fBatlc\fR - an Arbitrary Transmission Line Calculator
.SH SYNOPSIS
\fBatlc  [-C] [-s] [-S] [-v] [-c cutoff] [-d rrggbb=Er] [-i factor] [-t threads] [-r rate_multiplier] bitmapfile\fR
.br
.SH WARNING
This man page is not a complete set of documentation. See the html files
for more complete information. So far, I've not managed to install the
html files into /usr/local, so you will have to look into the
atlc-X.Y.Z/docs/html-docs directory for them. 
.SH DESCRIPTION
\fBatlc\fR is a finite difference programme that is used to calculate the 
properties of a two-conductor electrical transmission line of arbitrary 
cross section. It is used whenever there are no analytical formula known, 
yet you still require an answer. It can calculate:
.br
   The impedance Zo  (in Ohms)
.br
   The capacitance per unit length (pF/m)
.br
   The inductance per unit length (nF/m)
.br
   The velocity of propogation v (m/s)
.br
   The velocity factor, v/c, which is dimensionless. 
.PP 
A bitmap file (usually with the extension .bmp or .BMP) is drawn in a graphics 
package such as \fBGimp\fR available from http://www.gimp.org. The bitmap 
file \fBmust\fR be saved as a 24-bit (16 million colour) uncompressed file. The 
colours used in the bitmap indicate whether the region is a conductor (pure 
red, pure green or pure blue) or a dielectric (anything else). Pure white 
is assumed to be a vacuum dielectric, but other colours have diferent meanings. 
See COLOURS below for precise definitions of the colours. 
.SH OPTIONS
\fB-C\fR
.br
print copyright, licencing and copying information. 
.br
\fB-s\fR
.br
Skip writing the Ex, Ey, E, V, U and Er bitmap (.bmp) files 
\fB-S\fR
.br
Skip writing the Ex, Ey, E, V, U and Er binary (.bin) files 
.br
\fB-v\fR
.br
makes the output more verbose/talkative.
.br
\fB-c cutoff\fR
.br
Sets the convergence criteria of the finite difference programme. The
default is 0.0001, meaning two separate iterations must be within
01% for the programme to stop iterating. Setting to a smaller positive
number gives more accuracy, but takes longer. 
.PP
\fB-d rrggbb=Er\fR
.br
is used to indicate the colour 0xrrggbb in the bitmap is used to
represent a material with permittivity Er. See also COLOURS below
.PP
\fB-i factor\fR
.br
is used to lighten or darken the .bmp electric field profile images
produced by atlc. Set factor > 2 to lighten or between 1 and 2 to
darken.
.PP
\fB-r ratemultiplier\fR
.br
Sets the parameter 'r' used internally when computing the voltage at a point w,h.
The default, which is (as of version 3.0.0) 1.95, results in what is belived to be
optimal results. Setting to 1.0 will avoid the use of the fast convergence method,
which is generally not a good idea. 
.PP
\fB-t threads\fR
.br
tells atlc to run using that number of threads, to speed calculation. This
option is only valid if atlc was compiled with the --with-threads option to
configure. The pthreads library (or similar) \fBmust\fR be installed in order to compile
with support for more than one CPU. Generally, setting the number of threads to the
number of cpus is a good idea, but it is worth experimenting. The number can be
changed permanently by editing the parameter MAX_THREADS in definitions.h. 
.PP
.br
.PP
.br
.br
.br
.SH COLOURS
The 24-bitmaps that atlc uses have 8 bits assigned 
to represent the amount of red, 8 for blue and 8 for green. Hence there are 
256 levels of red, green and blue, making a total of 256*256*256=16777216 colours. 
Every one of the possible i16777216 colours can be defined precisely by the stating the exact amount 
of red, green and blue, as in:
.PP
.br
red         = 255,000,000 or 0xff0000
.br
green       = 000,255,000 or 0x00ff00
.br
blue        = 000,000,255 or 0x0000ff
.br
black       = 000,000,000 or 0x000000
.br
white       = 255,255,255 or 0xffffff
.br
Brown       = 255,000,255 or 0xff00ff 
.br
gray        = 142,142,142 or 0x8e8e8e
.PP
Some colours, such as pink, terquiose, sandy, brown, gray etc may mean slightly
different things to different prople. This is not so with atlc, as the 
programme expects the colours below to be exactly defined as given. Whether 
you feel the colour is sandy or yellow is up to you, but if you use it in 
your bitmap, then it either needs to be a colour reconised by atlc, \fBor\fR 
you \fBmust\fR define it with a command line option (see OPTIONS).
.br
red    = 255,000,000 or 0xFF0000 is the live conductor. 
.br
green  = 000,255,000 or 0x00FF00 is the grounded conductor. 
.br
blue   = 000,000,255 or 0x0000FF is the negative conductor 
.PP
All bitmaps \fBmust\fR have the live (red) and grounded (green) conductor. The blue
conductor is used to indicate a negative conductor, is needed when the programme 
is used to analyse directional couplers. 
.PP
The following dielectrics are recognised by \fRatlc\fB:
.PP
.br
white     255,255,255 or 0xFFFFFF as Er=1.0    (vacuum)
.br
pink      255,202,202 or 0xFFCACA as Er=1.0006 (air)
.br
L. blue   130,052,255 or 0x8235EF as Er=2.1    (PTFE)
.br
Mid gray  142,242,142 or 0x8E8E8E as Er=2.2    (duroid 5880)
.br
mauve     255.000,255 or 0xFF00FF as Er=2.33  (polyethylene)
.br
yellow    255,255,000 or 0xFFFF00 as Er=2.5    (polystyrene)
.br
sandy     239,203,027 or 0xEFCC1A as Er=3.3    (PVC)
.br
brown     188,127,096 or 0xBC7F60 as Er=3.335  (epoxy resin)
.br
L. yellow 223,247,136 or 0xDFF788 as Er=3.7    (FR4 PCB)
.br
Terquoise 026,239,179 or 0x1AEFB3 as Er=4.8    (glass PCB)
.br
Dark gray 142,142,142 or 0x696969 as Er=6.15   (duroid 6006)
.br
L. gray   240,240,240 or 0xDCDCDC as Er=10.2  (duroid 6010)
.br
D. orange 213,160,067 or 0xD5A04D as Er=100.0 (mainly for test purposes)
.br
.br
.SH EXAMPLES
Here are a few examples of the use of atlc. Again, see the html documentation in atlc-X.Y.Z\docs\html-docs
for for examples.     
.PP
\fBex_1 % atlc coax2.bmp\fR
.br
This is a simple example (ex_1), in which the geometry of a transmission line is defined in coax2.bmp. In
this example, only the predefined dielectrics (Er =1.0, 1.0006, 2.1, 2.2, 2.33, 2.5, 3.3, 3.335, 3.7,
4.8, 6.15
or 10.2) could have been used in the bitmap, which would have been done with one of 13 different colours. white (0xFFFFFF) for Er=1.0, pink (0xFFCACA) for 1.0006 etc.
No other colour(dilectric) could have been used, since it was not
specified with the -d option. 
.br
.PP
\fBex_2 % atlc -d f9e77d=2.43 somefile.bmp \fR 
.br
In ex_2, a dielectic with Er=2.43 was wanted. A colour with the RGB values of 0xF9E7&d was
used. The -d option tells atlc what Er this colour refers to.
.br
.PP
\fBex_3 % atlc -v .bmp coax2.bmp\fR 
.br 
In ex_3, atlc has been instructed to print the results of intermediate calculations to stdout.
Normally, only the final result is printed. 
.br
.SH FILES
bitmapfile.bmp
   Original bitmap file. Must be 24-bit colour uncompressed.
.br
bitmapfile.Ex.bmp
.br
   X-component of E-field as a bitmap. Red=+dV/dx, blue =-dV/dx
.br 
bitmapfile.Ey.bmp
   y-component of E-field as a bitmap. Red=+y, blue =-y
.br
bitmapfile.E.bmp
.br 
   E-field, as E=sqrt(Ex^2+Ey^2).
.br
bitmapfile.V.bin
.br 
   Voltage as a bitmap, red= positive, blue =negative.
.br
bitmapfile.Er.bin
.br
   Bitmap showing the permittivity as a grayscale. Lighter is a higher
   permittivity. 
.br 
bitmapfile.U.bmp
.br 
   Energy. 
.br
.P
In addition to the bitmaps, the data is also saved in binary files. 
.PP
All the saved binary files (.bin's) are saved as a double precision
number for each of the pixels. The first double is the top left, the
last the bottom right. If the original image has width W and height H,
the saved binary files will be W-1 by H-1. 
.PP
All the saved bitmap files are 24-bit uncompressed, just like the input
files. 
.
.SH SEE ALSO
atlc(1)
create_bmp_for_circ_in_circ(1)
create_bmp_for_circ_in_rect(1)
create_bmp_for_microstrip_coupler(1)
create_bmp_for_rect_cen_in_rect(1)
create_bmp_for_rect_cen_in_rect_coupler(1)
create_bmp_for_rect_in_circ(1)
create_bmp_for_rect_in_rect(1)
create_bmp_for_stripline_coupler(1)
create_bmp_for_symmetrical_stripline(1)
design_coupler(1)
find_optimal_dimensions_for_microstrip_coupler(1)
readbin(1)
.P 
.br
http://atlc.sourceforge.net                - Home page 
.br
http://sourceforge.net/projects/atlc       - Download area
.br
atlc-X.Y.Z/docs/html-docs/index.html       - HTML docs
.br
atlc-X.Y.Z/docs/qex-december-1996/atlc.pdf - theory paper
.br
atlc-X.Y.Z/examples                        - examples
.br
http://www.david-kirkby.co.uk              - my home page
.br
http://www.david-kirkby.co.uk/ham          - ham radio pages
.br


NEWS
NEWS for realease 4.3.0 on 23 May 2003.

A bug, preventing atlc from runningt properly on multiple
processors has been fixed. This is the main change. 

atlc should no longer fail any test on a multi-processor
system configured with --with-threads. The algorithm used
for both single processor and multiple processor (i.e. 
threaded code is the same). 

The -t option to atlc, which sets the number of threads 
when configured for multiple processors can be set to 0
to use entirely the single-threaded algorithm. If set to
one, it will use the multi-threaded algoritm, but use
only one thread. If set to some other number, it will use
that number of threads and be optimal for the same number
of cpus. 

Support for distributed programming has been temporarily removed.
Someone else added the code, but since he is no longer involved in
the development, it is difficult for me to follow. So I'll remove
the MPI specific code and will re-implement this. 

A test has been added to make sure atlc can handle the -d option 


Dr. David Kirkby 23/5/2003



src/atlc.c
#ifdef HAVE_STRING_H
#include <string.h>
#endif

#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif  

#include "definitions.h"

#ifdef ENABLE_MP

#ifdef HAVE_THREAD_H
#include <thread.h>
#endif

#ifdef HAVE_PTHREAD_H
#include <pthread.h>
#endif

#ifdef HAVE_PTHREADS_H
#include <pthreads.h>
#endif

#endif


#ifdef WINDOWS
#pragma hrdstop
#include <condefs.h>
#endif
struct pixels Er_on_command_line[MAX_DIFFERENT_PERMITTIVITIES];
struct pixels Er_in_bitmap[MAX_DIFFERENT_PERMITTIVITIES];

double **Vij;
double **Er;
double image_fiddle_factor;
int **cell_type; 
unsigned char *image_data;
int width, height;
extern int errno;
double found_this_dielectric=1000000.0;
int avoid_use_of_fast_convergence_method=FALSE;
size_t size;
int max_threads=MAX_THREADS, non_vacuum_found=FALSE;
int append_flag=FALSE;
int verbose=FALSE;
int dielectrics_to_consider_just_now;
int coupler=FALSE;
double r=1.95;

extern int main(int argc, char **argv) /* Read parameters from command line */
{
  long i;
  int q, offset, dielectrics_on_command_line=0, dielectrics_in_bitmap=0;
  int write_field_images=1;
  double capacitance, inductance, Zo, velocity, vf;
  double cutoff=0.0001, c, Zodd, Zeven;
  char *end, *output_filename, *input_filename;
  char *appendfile;
  FILE *image_data_fp, *appendfile_fp, *resultfile_fp;

  image_fiddle_factor=IMAGE_FIDDLE_FACTOR;
  end=string(0,1000);
  output_filename=string(0,1000);
  appendfile=string(0,1000);
  input_filename=string(0,1000);
  /* only use this if we have both a multi-threaded application and that 
  with have the function */
  while((q=get_options(argc,argv,"r:vsc:d:F:i:t:")) != -1)
  switch (q) 
  {
    case 'a':
      avoid_use_of_fast_convergence_method=TRUE;
    break;

    case 'd':
      /* Read a colour from the command line */
      Er_on_command_line[dielectrics_on_command_line].other_colour=\
      strtol(my_optarg, &end, 16);
      /* Sepparte it into the Red, Green and Blue components */
      Er_on_command_line[dielectrics_on_command_line].blue=\
      Er_on_command_line[dielectrics_on_command_line].other_colour%256;
      Er_on_command_line[dielectrics_on_command_line].green=\
      Er_on_command_line[dielectrics_on_command_line].other_colour/(256);
      Er_on_command_line[dielectrics_on_command_line].red=\
      Er_on_command_line[dielectrics_on_command_line].other_colour/(256*256);
      *end++; /* Gets rid of '=' sign which we put on the command line */
      Er_on_command_line[dielectrics_on_command_line].epsilon=atof(end);
      if (verbose)
        printf("r=%x g=%x b=%x col=%x Er=%f\n",\
      Er_on_command_line[dielectrics_on_command_line].red,\
      Er_on_command_line[dielectrics_on_command_line].green, \
      Er_on_command_line[dielectrics_on_command_line].blue, \
      Er_on_command_line[dielectrics_on_command_line].other_colour, \
      Er_on_command_line[dielectrics_on_command_line].epsilon);
      dielectrics_on_command_line++;
    break;
    case 'c':
      cutoff=atof(my_optarg);
    break;
    case 'r':
      r=atof(my_optarg);
    break;
    case 's':
      write_field_images=0;
    break;
    case 'F':
      strcpy(appendfile,my_optarg);
      append_flag=TRUE;
    break;
    case 't':
      max_threads=atol(my_optarg);
#ifndef ENABLE_MP
      error_and_exit("Error #1. The -t option can not be used on a version of \
atlc that was not\nconfigured with the --with-mp option, and hence built \
without the threads\nlibrary.\n",1);
#endif
    break;
      case 'i': /* Lighten or darken images of E-field */
      image_fiddle_factor=atof(my_optarg);
      if(image_fiddle_factor < 1.0)
      {
        fprintf(stderr,"Error #2.image_fiddle_factor set to %f. It must not be less than 1.\n",image_fiddle_factor);
        exit(2);
      }
    break;
    case 'v':
      verbose=TRUE;
    break;
    case '?':
      usage_atlc();
  } /* End of the switch statement */

  /* There should only be one argument to atlc, the bitmapfile name.
  There can be a few options though. We now check that there is only
  one argument */

  if(argc-my_optind == 1)  /* This should be so hopefully !! */
  {
    strcpy(input_filename, argv[my_optind]);
    strcpy(output_filename, input_filename);
    resultfile_fp=get_file_pointer_with_right_filename(input_filename,".txt");
    read_bitmap_file_headers(input_filename, &offset, &size, &width, &height);
    image_data=ustring(0L,size);
    /* On Solaris systems, if the following is not executed, only one 
    thread will run at any one time, which rather defeats the object of 
    running multi-threaded. */
#ifdef ENABLE_MP
#ifdef HAVE_THR_SETCONCURRENCY
    thr_setconcurrency(max_threads);
#endif
#endif 
    
    /* Each thread solves the equations for a part of the voltage
    matrix. If there were more threads than columms we would have a
    problem. I'm not sure exactly how many can be got away with, but
    one is unlikly to have more cpus that matrix columns */

    if(max_threads > width/4)
      max_threads=width/4;

    /* In theory, it would be sensible to get atlc to be able to read
    from stdin. This is a test, that checks if the filename is '-', and
    if so assumes its reading from stdin. So far, the programme is
    unable to read from stdin, so this code is not really doing
    anything useful, but might be expanded at a later date. */
    if( strcmp(argv[my_optind],"-") != 0)
    {
      if( (image_data_fp=fopen(output_filename, "rb")) == NULL)
      {
        fprintf(stderr,"Error #3. Can't open %s!!!!!\n", argv[my_optind]);
        exit(3);
      }
      if(fseek(image_data_fp, offset, SEEK_SET)!=0)
      {   
        fprintf(stderr,"Error #4. Sorry, unable to seek to the correct part \
	(offset=%d) of the bmp image\n", offset);
	exit(4);
      }
    }
    else
      image_data_fp=stdin;

    for(i=0; (i < (long)size ) && (feof(image_data_fp)==0); i++)
    {
      image_data[i]=(unsigned char)fgetc(image_data_fp);
    }
    if((ferror(image_data_fp) || (i != size)))
    {
      fprintf(stderr,"Error #5. Unable to read all of the image data properly\n");
      exit(5);
    }
    /* declare matrix's to indicate what pixels are fixed and variable */
    cell_type=imatrix(0,width,0,height);
    Vij=dmatrix(0,width,0,height);
    Er=dmatrix(0,width,0,height);
      
    /* We now fill the following 3 arrays with the correct data, based on the 
    contents of the bitmap image */

    setup_arrays(&dielectrics_in_bitmap, dielectrics_on_command_line);
    check_for_boundaries();
    /* Now 'v' has the voltages at each grid element, 'Er' 
    the permittivites and 'cell_type' is either FIXED or variable, 
    indicating a dielectric (VARIABLE) or a metal */

    /* If there are multiple dielectrics, the impedance calculations
    needs to be done twice. We start by doing them once, for an vacuum
    dielectric. If necessary, they will be done again */

    dielectrics_to_consider_just_now=1;
    if(coupler==FALSE)
      do_fd_calculation(&capacitance, &inductance, &Zo, &Zodd, &Zeven, Z0, \
      &velocity, &vf, stdout, cutoff, dielectrics_to_consider_just_now, \
      argv[my_optind],REQUIRE_FD_CALCULATIONS);
    else if(coupler==TRUE)
    {
      printf("**WARNING** This has a negative conductor and is therefore \
      \nconsidered in atlc as a coupler. This is highly experimental and \
      \nexpected to be suspect\n");

      do_fd_calculation(&capacitance, &inductance, &Zo, &Zodd, &Zeven, Z_ODD, \
      &velocity, &vf, stdout, cutoff, dielectrics_to_consider_just_now, \
      argv[my_optind],REQUIRE_FD_CALCULATIONS);

      swap_conductor_voltages(NEG_TO_POS);

      do_fd_calculation(&capacitance, &inductance, &Zo, &Zodd, &Zeven, Z_ALL, \
      &velocity, &vf, stdout, cutoff, dielectrics_to_consider_just_now, \
      argv[my_optind],REQUIRE_FD_CALCULATIONS);

      //swap_conductor_voltages(POS_TO_NEG);
    }
    /* The calculation of inductance above is correct, and does not
    need to be altered. However, if there are multiple dielectrics,
    then the capacitance needs to be computed again, this time taking
    into account the different dielectrics. If it just a single
    dielectric (like on a coax cable), then there is no need to repeat
    a lenghty FD calculation, but just to scale C by the relative
    permittivity of the dielectric. */

    if(dielectrics_in_bitmap > 1) /* Only do if there is a mixed dielectric */
    {
      dielectrics_to_consider_just_now=dielectrics_in_bitmap;
      do_fd_calculation(&capacitance, &inductance, &Zo, &Zodd, &Zeven,Z0 , &velocity, &vf, stdout, cutoff, dielectrics_to_consider_just_now, argv[my_optind],REQUIRE_FD_CALCULATIONS);

    }
    /* Now we have findished all the time-consuming Fintite difference 
    bits */

    /* If there is more than one dielectric, but it is not vacuum, then
    the original calculation of C assuming vacuum just needs scaling by
    the relative permittivity of the dielectric. Very easy to do 
    and there is *NO* need for any more finite difference calculations*/
#ifndef GG
    if((dielectrics_in_bitmap == 1) && (non_vacuum_found==TRUE))
    {
      capacitance*=found_this_dielectric;
      /* inductance is alreadly correct */
      Zo=sqrt(inductance/capacitance);
      velocity=1.0/pow(inductance*capacitance,0.5);
      c=1.0/(sqrt(MU_0 * EPSILON_0));
      vf=velocity/c;
      if(verbose)
      {
        print_data(stdout,argv[my_optind],found_this_dielectric,\
	capacitance,inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
        if(append_flag==TRUE)
        {
          appendfile_fp=fopen(appendfile,"a");
	  if(appendfile_fp==NULL)
	  {
	     fprintf(stderr,"Error #8. Can't open file for appending data in atlc.c\n");
	     exit(8);
          }
	  print_data(appendfile_fp,argv[my_optind],found_this_dielectric,\
	  capacitance,inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
          fclose(appendfile_fp);
        }
      }
    } /* end of checking for a single dielectric, that is not vacuum */
#endif
    if(dielectrics_in_bitmap > 1)
    {
      if(coupler==FALSE)
      {
        print_data(stdout,argv[my_optind],-1.0,capacitance,\
        inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
        print_data(resultfile_fp,argv[my_optind],-1.0,capacitance,\
        inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
      }
      else if (coupler==TRUE)
      {
        print_data(stdout,argv[my_optind],-1.0,capacitance,\
        inductance,Zo,Zodd,Zeven,Z_ALL,velocity,vf);
        print_data(resultfile_fp,argv[my_optind],-1.0,capacitance,\
        inductance,Zo,Zodd,Zeven,Z_ALL,velocity,vf);
      }
      if(append_flag==TRUE)
      {
        appendfile_fp=fopen(appendfile,"a");
        print_data(appendfile_fp,argv[my_optind],-1.0,capacitance,\
	inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
            fclose(appendfile_fp);
	  }
      }
      else if((dielectrics_in_bitmap == 1) && (non_vacuum_found==TRUE))
      {
         print_data(stdout,argv[my_optind],found_this_dielectric,capacitance,\
	 inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
         print_data(resultfile_fp,argv[my_optind],found_this_dielectric,\
	 capacitance,inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
	 if(append_flag==TRUE)
	 {
	   appendfile_fp=fopen(appendfile,"a");
           print_data(appendfile_fp,argv[my_optind],found_this_dielectric,\
	   capacitance,inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
           fclose(appendfile_fp);
	 }
      }
      else
      {
        if(coupler==FALSE)
        {
          print_data(stdout,argv[my_optind],1.0,capacitance,inductance,\
	  Zo,Zodd,Zeven,Z0,velocity,vf);

          print_data(resultfile_fp,argv[my_optind],1.0,capacitance,\
	  inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
        }
	else if(coupler==TRUE)
	{
          print_data(stdout,argv[my_optind],1.0,capacitance,inductance,\
	  Zo,Zodd,Zeven,Z_ALL,velocity,vf);

          print_data(resultfile_fp,argv[my_optind],1.0,capacitance,\
	  inductance,Zo,Zodd,Zeven,Z_ALL,velocity,vf);
        }
	if(append_flag==TRUE && coupler==FALSE)
	{
	   appendfile_fp=fopen(appendfile,"a");
           print_data(appendfile_fp,argv[my_optind],1.0,capacitance,\
	   inductance,Zo,Zodd,Zeven,Z0,velocity,vf);
           fclose(appendfile_fp);
	}
	else if(append_flag==TRUE && coupler==TRUE)
	{
	   appendfile_fp=fopen(appendfile,"a");
           print_data(appendfile_fp,argv[my_optind],1.0,capacitance,\
	   inductance,Zo,Zodd,Zeven,Z_ALL,velocity,vf);
           fclose(appendfile_fp);
	}
      }
      /* The function write_fields needs to have a file with the
      extension .bmp. We need to be sure this is the case, so if not, we
      append .bmp
      
      Currently the fields on copulers are not properly defined
      so I won't plot them. */
      
      if(write_field_images==TRUE && coupler==FALSE)
        write_fields(output_filename);
      free_dmatrix(Vij,0,width,0,height);
      free_dmatrix(Er,0,width,0,height);
      exit(0);
   }
   else
      usage_atlc();
   exit(5);
}
